cmds/chmod/chmod_test.go

```diff
@@ -0,0 +1,78 @@
+package main
+
+import (
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+var (
+	testPath = "."
+	// if true removeAll the testPath on the end
+	remove = true
+)
+
+const (
+	maxSizeFile = 4096
+	rangeRand   = 100
+	maxDirDepth = 5
+	maxFiles    = 5
+)
+
+// randomFile create a random file with random content
+func randomFile(fpath, filename string) (*os.File, error) {
+	f, err := ioutil.TempFile(fpath, filename)
+	if err != nil {
+		return nil, err
+	}
+	// generate random content for files
+	filesize := rand.Intn(maxSizeFile)
+	bytes := make([]byte, filesize)
+	rand.Read(bytes)
+	f.Write(bytes)
+
+	return f, nil
+}
+
+func TestChmodSimple(t *testing.T) {
+	tempDir, err := ioutil.TempDir("", "TestChmodSimple")
+	if err != nil {
+		t.Fatalf("cannot create temporary directory: %v", err)
+	}
+	if remove {
+		defer os.RemoveAll(tempDir)
+	}
+	testpath := filepath.Join(tempDir, "testchmod.exe")
+	f, err := randomFile(tempDir, "BLAH1")
+	if err != nil {
+		t.Fatalf("cannot create temporary file: %v", err)
+	}
+	defer f.Close()
+
+	// Build
+	out, err := exec.Command("go", "build", "-o", testpath, ".").CombinedOutput()
+	if err != nil {
+		t.Fatalf("go build -o %v cmds/chmod: %v\n%s", testpath, err, string(out))
+	}
+
+	// Set permissions and check that they are correct
+	c := exec.Command(testpath, "0777", f.Name())
+	c.Run()
+	fileinfo, _ := os.Stat(f.Name())
+	fileperm := fileinfo.Mode().Perm()
+	if fileperm != 0777 {
+		t.Errorf("Wrong file permissions")
+	}
+
+	// Change permissions and check that they are correct
+	c = exec.Command(testpath, "0644", f.Name())
+	c.Run()
+	fileinfo, _ = os.Stat(f.Name())
+	fileperm = fileinfo.Mode().Perm()
+	if fileperm != 0644 {
+		t.Errorf("Wrong file permissions")
+	}
```

Comment:
It's probably worth adding incorrect invocations too, e.g.
chmod 1777
chmod abdc
and so on. This can be done with a table